pkg/securelaunch/policy/policy.go

```diff
@@ -0,0 +1,117 @@
+package policy
+
+import (
+	"encoding/json"
+	"errors"
+	"log"
+
+	"github.com/u-root/u-root/pkg/diskboot"
+	"github.com/u-root/u-root/pkg/mount"
+	slaunch "github.com/u-root/u-root/pkg/securelaunch"
+	"github.com/u-root/u-root/pkg/securelaunch/eventlog"
+	"github.com/u-root/u-root/pkg/securelaunch/launcher"
+	"github.com/u-root/u-root/pkg/securelaunch/measurement"
+)
+
+/*
+ * policy structure passed around various sub-packages (
+ * measurement, eventlog and launcher) in securelaunch.
+ * Used to store parsed policy file. The policy file
+ * is a JSON file.
+ */
+type policy struct {
+	DefaultAction string
+	Collectors    []measurement.Collector
+	Launcher      launcher.Launcher
+	EventLog      eventlog.EventLog
+}
+
+/*
+ * Locate() ([]byte, error)
+ * Check if kernel param sl_policy is set,
+ *	- parse the string
+ * Iterate through each local block device,
+ *	- mount the block device
+ *	- scan for securelaunch.policy under /, /efi, or /boot
+ * Read in policy file
+ */
+func Locate() ([]byte, error) {
+
+	d, _ := slaunch.ScanKernelCmdLine()
+	if d != nil {
+		return d, nil
+	}
+
+	slaunch.Debug("Searching and mounting block devices with bootable configs\n")
+	blkDevices := diskboot.FindDevices("/sys/class/block/*") // FindDevices find and *mounts* the devices.
+	if len(blkDevices) == 0 {
+		return nil, errors.New("No block devices found")
+	}
+
+	for _, device := range blkDevices {
+		devicePath, mountPath := device.DevPath, device.MountPath
+		slaunch.Debug("scanning for policy file under devicePath=%s, mountPath=%s\n", devicePath, mountPath)
+		raw, found := slaunch.ScanBlockDevice(mountPath)
+		if e := mount.Unmount(mountPath, true, false); e != nil {
+			log.Printf("Unmount failed. PANIC\n")
+			panic(e)
+		}
+
+		if !found {
+			log.Printf("no policy file found under this device\n")
+			continue
+		}
+
+		slaunch.Debug("policy file found at devicePath=%s\n", devicePath)
+		return raw, nil
+	}
+
+	return nil, errors.New("policy file not found anywhere.")
+}
+
+/*
+ * Parse(pf []byte) (*policy, error)
+ * Sets up policy structure by unmarshalling policy file.
+ * returns pointer to policy structure.
+ */
+func Parse(pf []byte) (*policy, error) {
```

Comment:
policy should be exported -- an exported function should only return exported (= publicly available) types